### PR TITLE
[SYCL] Predefine SYCL macros with vendor strings

### PR DESCRIPTION
**Pull request by dsawfer**
The SYCL 2020 specification requires at least one macro of the form SYCL_IMPLEMENTATION_"vendorstring" to be predefined, which allows applications to check if they compile with this provider's implementation.

Our vendorstrings are SYCL_IMPLEMENTATION_ONEAPI and SYCL_IMPLEMENTATION_INTEL. The SYCL_FEATURE_SET_FULL macro is mandated by section 5.6. "Preprocessor directives and macros" of the SYCL 2020 specification.

### CODE REVIEW

File: sycl/test/basic_tests/define_vendors.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
```
Comment (alexbatashev) [inline sycl/test/basic_tests/define_vendors.cpp:1]:
```suggestion
// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -c -o %t.out
```

File: sycl/test/basic_tests/define_vendors.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+#if !defined(SYCL_FEATURE_SET_FULL)
+#error SYCL_FEATURE_SET_FULL is not defined
+#endif
+
+#if !defined(SYCL_IMPLEMENTATION_INTEL)
+#error SYCL_IMPLEMENTATION_INTEL is not defined
+#endif
+
+int main(void) { return 0; }
```
Comment (alexbatashev) [inline sycl/test/basic_tests/define_vendors.cpp:16]:
```suggestion
```

File: sycl/include/CL/sycl/define_vendors.hpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#define SYCL_IMPLEMENTATION_ONEAPI
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/define_vendors.hpp:11]: Could you please clarify why we define this macro?

Comment (alexbatashev) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: @romanovvlad https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_preprocessor_directives_and_macros

Comment (romanovvlad) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: Do not see anything about defining `SYCL_IMPLEMENTATION_ONEAPI` in section 5.6.
Could you please copy paste specific wording which requires us to define this macro?

Comment (alexbatashev) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: @romanovvlad sorry, I've forgot to put the link the first time I posted the comment. I've updated it with the link to  section 6.2, which reads:

> The implementation must predefine at least one macro of the form SYCL_IMPLEMENTATION_<vendorstring> which allows applications to test whether they are being compiled with that vendor’s implementation. For example, the Acme vendor could predefine a macro whose name is SYCL_IMPLEMENTATION_ACME.

Comment (romanovvlad) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: Thanks, it sounds like `SYCL_IMPLEMENTATION_INTEL` covers that. Could you please clarify why `SYCL_IMPLEMENTATION_ONEAPI` is being added?

File: sycl/include/CL/sycl/define_vendors.hpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+#define SYCL_FEATURE_SET_FULL
+#define SYCL_IMPLEMENTATION_INTEL
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/define_vendors.hpp:13]: I believe we need to document the macro in https://github.com/intel/llvm/blob/sycl/sycl/doc/PreprocessorMacros.md

File: sycl/include/CL/sycl/define_vendors.hpp (hunk 3 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: Discussed offline.

Comment (alexbatashev) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: We provide extensions in both `ext::oneapi` and `ext::intel` namespaces. My understanding is that we must provide both macros as well. Tagging @gmlueck

Comment (gmlueck) [reply sycl/include/CL/sycl/define_vendors.hpp:11]: Agreed, we should define both.

File: sycl/include/CL/sycl/define_vendors.hpp (hunk 4 above)
Comment (gmlueck) [reply sycl/include/CL/sycl/define_vendors.hpp:13]: I'm not sure about this.  I thought that document described the preprocessor macros that the **user** can set, which have an effect on the compiler.  However, the user is not supposed to set macros like `SYCL_IMPLEMENTATION_INTEL`.  Rather, the implementation sets the macros, and these user can only test them.

**Review by gmlueck — APPROVED** [review]

**Review by alexbatashev — APPROVED** [review]

File: sycl/include/CL/sycl/define_vendors.hpp (hunk 4 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/define_vendors.hpp:13]: I think we should have two documents: for macros, that user defines, and for macros, that compiler defines, but that’s a job for a separate PR.

**Comment by dm-vodopyanov** [comment]
@dsawfer @alexbatashev Should tests need to be added?

**Comment by alexbatashev** [comment]
> @dsawfer @alexbatashev Should tests need to be added?

There's a test in the tests folder. What other kind of testing are you thinking of?

**Comment by dm-vodopyanov** [comment]
> > @dsawfer @alexbatashev Should tests need to be added?
> 
> There's a test in the tests folder. What other kind of testing are you thinking of?

Sorry, missed that somehow.

**Comment by dm-vodopyanov** [comment]
> @dsawfer, can you please add a description for this PR?

Thanks!

**Pull request merged**